### PR TITLE
fix: decode path before matching it

### PR DESCRIPTION
**Pull request by rainum**
**Summary**

You have a spec with `/some path` in it and you're trying to mock this path with mocker. If you'll send an encoded path (`/some%20path`) to the mocker it wont be able to find a corresponding path in the spec at will throw `NO_PATH_MATCHED_ERROR` which is not correct.

**Checklist**

- The basics
  - [x] I tested these changes manually in my local or dev environment
- Tests
  - [x] Added or updated
  - [ ] N/A
- Event Tracking
  - [ ] I added event tracking and followed the event tracking guidelines
  - [x] N/A
- Error Reporting
  - [ ] I reported errors and followed the error reporting guidelines
  - [x] N/A

### CODE REVIEW

File: packages/http/src/router/matchPath.ts [hunk 1]
```diff
@@ -1,18 +1,29 @@
 import { MatchType } from './types';
 import * as E from 'fp-ts/Either';
 
-function fragmentarize(path: string): string[] {
+function fragmentize(path: string): string[] {
```
Comment (rainum) [inline packages/http/src/router/matchPath.ts:4]: I decided to update working a bit

File: packages/http/src/router/matchPath.ts [hunk 2]
```diff
@@ -1,18 +1,29 @@
...
-function getTemplateParamName(pathFragment: string) {
-  const match = /{(.*)}/.exec(pathFragment);
+// Attempt to decode path fragment. Decode should not do any harm since it is
+// decoding only URI sequences which are previously created by encodeURIComponent
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent
+function decodePathFragment(pathFragment?: string) {
+  try {
+    return pathFragment && decodeURIComponent(pathFragment);
+  } catch (_) {
+    return pathFragment;
```
Comment (radzserg) [inline packages/http/src/router/matchPath.ts:15]: when `decodeURIComponent` can throw an error?

File: packages/http/src/router/matchPath.ts [hunk 3]
```diff
@@ -1,18 +1,29 @@
...
+  }
+}
+
+function getTemplateParamName(pathFragment?: string) {
+  const match = typeof pathFragment === 'string' && /{(.*)}/.exec(pathFragment);
```
Comment (radzserg) [inline packages/http/src/router/matchPath.ts:20]: I feel like `mathPath` lacks a functional approach. How to turn 2 linkes into 11 lines ;)

```
  return pipe(
    O.fromNullable(pathFragment),
    O.map(pathFragment =>
      pipe(
        pathFragment,
        O.fromPredicate(pathFragment => typeof pathFragment === 'string'),
        O.map((pathFragment: string) => /{(.*)}/.exec(pathFragment)),
        O.map(m => m && m[1])
      )
    )
  );

```

File: packages/http/src/router/matchPath.ts (hunk 2 above)
Comment (rainum) [reply packages/http/src/router/matchPath.ts:15]: It'll throw if you provide any invalid sequence, for example: `decodeURIComponent('%1')`. I just want to make sure that we can handle some weird URLs.

File: packages/http/src/router/matchPath.ts (hunk 3 above)
Comment (rainum) [reply packages/http/src/router/matchPath.ts:20]: Yeah, it definitely is. In this particular case I just fixed type issue, let's plan a functional overhaul of `matchPath.ts` for another time 😉

**Review by radzserg — APPROVED** [review]
LGTM 👍

**Pull request merged**